sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/PartitionBasedLoadBalancer.java

```diff
@@ -0,0 +1,333 @@
+     */
+    private String findPartitionToSteal(final Map<String, List<PartitionOwnership>> ownerPartitionMap) {
+        List<PartitionOwnership> maxList = ownerPartitionMap.values()
+            .stream()
+            .max(Comparator.comparingInt(List::size))
+            .get();
+
+        return maxList.get(RANDOM.nextInt(maxList.size())).partitionId();
+    }
+
+    /*
+     * When the load is balanced, all active event processors own at least {@code minPartitionsPerEventProcessor}
+     * and only {@code numberOfEventProcessorsWithAdditionalPartition} event processors will own 1 additional
+     * partition.
+     */
+    private boolean isLoadBalanced(final int minPartitionsPerEventProcessor,
+        final int numberOfEventProcessorsWithAdditionalPartition,
+        final Map<String, List<PartitionOwnership>> ownerPartitionMap) {
+
+        if (ownerPartitionMap.values()
+            .stream()
+            .noneMatch(ownershipList -> {
+                return ownershipList.size() < minPartitionsPerEventProcessor
+                    || ownershipList.size() > minPartitionsPerEventProcessor + 1;
+            })) {
+            long count = ownerPartitionMap.values()
+                .stream()
+                .filter(ownershipList -> ownershipList.size() == minPartitionsPerEventProcessor + 1)
+                .count();
+
+            return count == numberOfEventProcessorsWithAdditionalPartition;
+        }
+        return false;
+    }
+
+    /*
+     * This method is called after determining that the load is not balanced. This method will evaluate
+     * if the current event processor should own more partitions. Specifically, this method returns true if the
+     * current event processor owns less than the minimum number of partitions or if it owns the minimum number
+     * and no other event processor owns lesser number of partitions than this event processor.
+     */
+    private boolean shouldOwnMorePartitions(final int minPartitionsPerEventProcessor,
+        final Map<String, List<PartitionOwnership>> ownerPartitionMap) {
+
+        int numberOfPartitionsOwned = ownerPartitionMap.get(this.ownerId).size();
+
+        int leastPartitionsOwnedByAnyEventProcessor =
+            ownerPartitionMap.values().stream().min(Comparator.comparingInt(List::size)).get().size();
+
+        if (numberOfPartitionsOwned < minPartitionsPerEventProcessor
+            || numberOfPartitionsOwned == leastPartitionsOwnedByAnyEventProcessor) {
+            return true;
+        }
+        return false;
+    }
+
+    /*
+     * This method will create a new map of partition id and PartitionOwnership containing only those partitions
+     * that are actively owned. All entries in the original map returned by PartitionManager that haven't been
+     * modified for a duration of time greater than the allowed inactivity time limit are assumed to be owned by
+     * dead event processors. These will not be included in the map returned by this method.
```

Comment:
I wonder if we should add a tiny bit of random jitter to the time calculations.   If a bunch of event processors started running at just about the same time, then we're likely to see contention.  A bit of jitter to have them working at a slightly earlier or later duration for each calculation may help alleviate that.